Fb ufs regtests

aliabdolali opened the pull request:
# Pull Request Summary 
* Adding ww3_grib build and run within _regtests/bin/run_test_.
* Updating the required libraries on Hera and Orion using hpc-stack for ww3_grib in _regtests/bin/matrix_ncep_.
* Adding _regtests/bin/test.comp_ to compare two runs from the same code version for b4b reproducibility check.
* Reintegrate the Fix for arctic grib files #346 from https://github.com/NOAA-EMC/WW3/tree/production/GFS.v16
* Add the following regtests for ufs application.
* Separate the MPI and OMP/OMPH tests in matrix_divider_p.sh
1. global model with wind, ice, and current, taken from ufs setup
- adding the -C ESMF test for the 1 deg global model 
- adding the restart reproducibility check
- adding the MPI task reproducibility check
- adding the thread reproducibility check
- adding the ww3_grib test with NCO NCEP2 swtich
2. Adding GFSv16 setup with wind (hourly), ice (daily), and current (3hr interval)
- adding the -C ESMF test for the GFSv16
- adding the restart reproducibility check
- adding the ww3_grib test with NCO NCEP2 swtich

For regtests/bin/matrix_ncep, the required modules for esmf and grib outputs are added (hpc-stack).
The regtests/bin/matrix.base is updated and the check statement for restart b4b reproducibility is added. 
The regtests/bin/matrix_divider_p.sh is updated and the large ufs tests which requires a larger number of cpus are separated.


## issues
https://github.com/NOAA-EMC/WW3/issues/317
https://github.com/NOAA-EMC/WW3/issues/214
https://github.com/NOAA-EMC/WW3/issues/323
https://github.com/NOAA-EMC/WW3/issues/322
https://github.com/NOAA-EMC/WW3/issues/181
closed https://github.com/NOAA-EMC/WW3/issues/345
closed https://github.com/NOAA-EMC/WW3/issues/276
closed https://github.com/NOAA-EMC/WW3/issues/366
closed https://github.com/NOAA-EMC/WW3/issues/367
## Description
This development is for large setups specifically for UFS application (esmf) and GFSv16.


* Reviewers: @JessicaMeixner-NOAA 


### Testing
two regression tests are added
ww3_ufs1.1
ww3_ufs.1.2 (GFSv16)
There should be no changes in the regression test except the known non-identical cases and ww3_tp2.17 due to the changes in the inputs. 
* Which compiler / HPC you used to run the regression tests in the PR? 
Testes on Hera with hpc-stack libraries



An issue related to GFSv16 setup is added (see https://github.com/NOAA-EMC/WW3/issues/417)

UKMO-lsampson committed:
Added boundary checks to the SMC grid input files for ww3_grid,
to ensure they comply with the limits of the nameslist.

ukmo-juan-castillo committed:
Fb 360 calendar (#8)

Changes to add support to 360-day and 365-day (no leap year) calendar - see ticket #209
  * Additional CALTYPE namelist parameter in MISC section
  * New ww3_tc1 regtest.

ukmo-ccbunney committed:
RTD support for ww3_boun[dc] (#10)

* Updated ww3_bound and ww3_bounc to handle model grids formulated on a rotated pole.
* Manual and nml/inp files to updated clarify that ww3_bound/ww3_bounc only accept input spectra formulated on a standard pole grid.

ukmo-juan-castillo committed:
Fb coupling time (#9)

Updates to allow a coupling time step that is different from the model time step. 
* Includes new regtest (in ww3_tp2.14) for non-default oasis time step.
* ww3_tp2.14 regtest added to matrix.base.

Commit:
Merge remote-tracking branch 'upstream/staging' into develop

Commit:
bug fix for ukmet development

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
add global 1deg ufs sertup as stand alone ww3 test

Commit:
add nc file

Commit:
add more

Commit:
add more

aliabdolali committed:
Delete wind_cur_ice.nc

Commit:
rm extras

aliabdolali committed:
Delete regtests/ww3_esmf1.1/input directory

Commit:
modify divider for ufs tests

Commit:
add more stuff

Commit:
add test.comp script

Commit:
edit

Commit:
correct ww3_grib inputs

Commit:
correct ww3_grib inputs for ufs test

Commit:
modify run_test to generate multiple grib file

Commit:
no commit message

Commit:
change -o grib to -o all

Commit:
add info and threat and mpi task tests

Commit:
update ww3_from_ftp.sh with the new tar file on ftp and run_test update message

Commit:
add unstructured restart b4b check

Commit:
clean test.comp script

Commit:
fix bug in divider

Commit:
more clean up

Commit:
add missing ounf to ufs tests and fix a typo in info

Commit:
add gint to ww3_ufs1.2 and new grib for the integrated grid

Commit:
add ww3_multi and ww3_gint to ww3_tp2.21 (global unstr) test

Commit:
add ww3_tp2.21 new test to matrix.base

aliabdolali commented:
I added the issues this PR addresses (some are addressed partially, not closing the issues).
The matrix ran fine on NOAA HPC
See the summary of matrix.comp with expected non-b4b cases. The added tests ran fine and test.comp did the comparison for the same version of the code. 
[MatrixCOMP_UFS.zip](https://github.com/NOAA-EMC/WW3/files/6435204/MatrixCOMP_UFS.zip)

Commit:
update manual for new regtests

Commit:
edit typos in manual

JessicaMeixner-NOAA commented on regtests/bin/matrix_ncep:
```diff
@@ -179,7 +212,12 @@
   export  multi07='y' # mww3_test_07 (unstructured grid tests)
   export  multi08='y' # mww3_test_08 (wind and ice tests)
-
+  export      ufs='y' # The Unified Forecast System
```
Updates to matrix_ncep are not propagated to other matrix_machines.  Likely these should be turned to 'n' but the options should be propagated to all the matrix files.

aliabdolali replied:
thanks for noticing it. I'll add ufs='n' to all the matrix_?

aliabdolali replied:
Done

JessicaMeixner-NOAA commented on regtests/ww3_ufs1.1/input/grdset_a:
```diff
@@ -0,0 +1,3 @@
+MODEL: glo_1deg
+INPUT: glo_30m
```
Why not just do glo_1deg?

JessicaMeixner-NOAA replied:
Or even not having an input grid seems fine too?

aliabdolali replied:
These tests are checking ww3_multi with input grids, and I wanted to have it ~= the computational grids, so we can check all the functionality on this scale. I have taken the ice, current and wind from ufs-weather tests to be consistent with them. Of course these are not exactly the same as ufs-weather.
For ww3_gint and ww3_grib, I did the same, just to check if all functions/interpolation are working fine.

JessicaMeixner-NOAA replied:
Because of the other issues, can you confirm that the glo_1deg is fully covered by the glo_30m or should this also be revisited with the needed input change in ufs1.2?

JessicaMeixner-NOAA commented on regtests/ww3_ufs1.2/info:
```diff
@@ -0,0 +1,56 @@
+#   > Arctic (aoc_9km); Resolution: 9km; Lat. Range: 50N-90N.               #
+#   > Global core (gnh_10m); Resolution: 1/6°; Lat. Range: 15S-52N          #
+#   > Southern Ocean (gsh_15m); Resolution: 1/4°; Lat. Range: 79.5S-10.5S.  #
+# INPUT: glo_30m. The forcing including ice concentration, wind speed (u10) #
```
Why glo_30m forcing? This is much less than what we are using in GFSv16? Is the ice/current sourced from operations or something else?

JessicaMeixner-NOAA replied:
I think it's fine for this test, but if you plan on using this test as a surrogate for improvements in a GFSv16 set-up this will not be sufficient.

aliabdolali replied:
> Why glo_30m forcing? This is much less than what we are using in GFSv16? Is the ice/current sourced from operations or something else?

Yes, I'll update it later when I have optimization ready.

aliabdolali replied:
Also, note that #368 checks the ufs-weather inside ufs-weather-model changing the WW3 to the one under development.

JessicaMeixner-NOAA replied:
So another issue with using the glo_30m as the input grid is that it does not cover the entire globe.  Even with your previous commit that works (better), you still get this and so I think we need to revisit the input source and input grid for this case reguardless of what appears to be some other bugs and issues
<img width="499" alt="Screen Shot 2021-06-01 at 6 01 23 PM" src="https://user-images.githubusercontent.com/22549231/120395675-631b8c00-c303-11eb-96f5-0f44896a8904.png">

aliabdolali replied:
no, 1deg lat is [-85 +85] and 30 min is [-80 80]. So I'll replace the forcing that covers both ufs1.1 and ufs1.2.

JessicaMeixner-NOAA reviewed (commented):
This is my first round of review comments.  I still need to run the new tests, etc.  One thing I did notice is that if we add these ufs tests, they break the original matrix as the resource requests do not match.  Not sure if we care about that or not, but something to note that it no longer works without after running matrix_diver_p, which almost makes me think if we can run w/out the diver, why not just call matrix_divider_p at the end of matrix_ncep?

aliabdolali commented:
> This is my first round of review comments. I still need to run the new tests, etc. One thing I did notice is that if we add these ufs tests, they break the original matrix as the resource requests do not match. Not sure if we care about that or not, but something to note that it no longer works without after running matrix_diver_p, which almost makes me think if we can run w/out the diver, why not just call matrix_divider_p at the end of matrix_ncep?

We can add matrix_divider_p to the end of matrix_ncep or alternatively, we can set ufs='n' for all the machines (including ncep) as default and whenever we want to test it, manualy turn it to 'y' at ncep.

Commit:
update matrix_datarmor matrix_milhydro matrix_ukmo_cray

JessicaMeixner-NOAA commented on regtests/bin/matrix_ncep:
```diff
@@ -63,15 +71,26 @@
-  export  nr='4'   #number of mpi tasks for hybrid
-  export nth='6'   #number of threads
+  export  np='24'     #number of mpi tasks
+  export  npl='440'   #number of mpi tasks for ufs applications and large setups
```
440*8 seems excessive for a regression test.  While this is close to the operational set-up which might be important to test on a slower cadence than every commit, for GDAS we use 140 tasks and 4threads which seems like a much more manageable number, especially given that the 1degree ufs 1.1 is also using these task counts.     Is there a reason behind the choice of using --cpus-per-task=${nth} instead of just requesting a certain set of nodes?

aliabdolali replied:
I'll test the suggested combinations and will report.

aliabdolali replied:
@JessicaMeixner-NOAA Done

JessicaMeixner-NOAA commented on regtests/bin/matrix.base:
```diff
@@ -2063,7 +2084,83 @@
+  then
+    echo ' ' >> matrix.body
+    echo "mkdir -p ww3_ufs1.2/work_a_esmf"                                                                    >> matrix.body
+    echo "cp ww3_ufs1.2/input/rmp_src* ww3_ufs1.2/work_a_esmf/"                                               >> matrix.body
```
Should this be added to the run_test script instead of the matrix.base?

aliabdolali replied:
I followed the same logic for other tests that have dependencies to use mkdir in matrix.base.
I will move ` echo "cp ww3_ufs1.2/input/rmp_src* ww3_ufs1.2/work_a_esmf/"` to run_test. Good suggestion.

aliabdolali replied:
@JessicaMeixner-NOAA Done, now I am going to check the fields

JessicaMeixner-NOAA commented on regtests/bin/matrix_ncep:
```diff
@@ -179,7 +212,12 @@
+  export rstrt_b4b='y' # Restart Reproducibility
+  export  npl_b4b='y' # MPI task Reproducibility
+  export  nth_b4b='y' # Thread Reproducibility
+  export     esmf='n' # ESMF coupling
```
This is added but turned off? Is this holding a place for future updates?

aliabdolali replied:
Yes, with esmf='y', we could test the compilation of ww3_multi_esmf, but then I realized we cannot execute it. At first, I thought the regtests/bin/run_esmf_test_suite script that Tim had added was working, so I wanted to use the same functionality. It did not work as expenced. so I left it here for future development. Instead, I started #368 which seems an alternative solution. We can discuss it in a tag up.

JessicaMeixner-NOAA commented:
Running the matrix11 and matrix12 as a sanity check for the new additions and look at the output.  I made few comments, particularly about reducing the node count on the ufs jobs, but as long as these tests work this PR should be ready to go.

JessicaMeixner-NOAA commented:
@aliabdolali  I have run the matrix11 and matrix12 sections, to look at the new output and the test.comp for the ww3_tp2.17 test does not look at the actual netcdf file because it's in a folder, see output here:

```
       *****************************************************************
     ***  compare WAVEWATCH III ww3_tp2.17/work_ma and ww3_tp2.17/work_ma1  ***
       *****************************************************************

inla.grb2 is identical
mapsta.ww3 is identical
mask.ww3 is identical
nest.inla is identical
Warning: out_grd.inla is different
Warning: restart001.inla is different
Warning: restart002.inla is different
cmp: ww3_tp2.17/work_ma/ww3_ounf_inla: Is a directory
Warning: ww3_ounf_inla is different

   ****************************************************************************
 ***  end of WAVEWATCH III compare tests with the same version of the code   ***
   ****************************************************************************
```

Is this expected behavior? 


After the grib programs run, we are getting: 
warning: empty layer generated 
and in fact if you convert the grib files to netcdf, they are all empty.  This is likely a module file issue.  We need to use jasper/2.0.25 for the grib files to be non-empty.

JessicaMeixner-NOAA commented:
HS looks weird to me in ww3_ufs1.2/work_a, have you looked at that? Maybe it's just me?

aliabdolali commented:
> Running the matrix11 and matrix12 as a sanity check for the new additions and look at the output. I made few comments, particularly about reducing the node count on the ufs jobs, but as long as these tests work this PR should be ready to go.

Thanks for catching that. I'll add the functionality to compare subdirectories as well.
Also, I will change the jasper version and will test it.

Commit:
update

Commit:
update bin/matrix_divider_p.sh to separate OMP/OMPH from MPI tests

aliabdolali commented:
> @aliabdolali I have run the matrix11 and matrix12 sections, to look at the new output and the test.comp for the ww3_tp2.17 test does not look at the actual netcdf file because it's in a folder, see output here:
> 
> ```
>        *****************************************************************
>      ***  compare WAVEWATCH III ww3_tp2.17/work_ma and ww3_tp2.17/work_ma1  ***
>        *****************************************************************
> 
> inla.grb2 is identical
> mapsta.ww3 is identical
> mask.ww3 is identical
> nest.inla is identical
> Warning: out_grd.inla is different
> Warning: restart001.inla is different
> Warning: restart002.inla is different
> cmp: ww3_tp2.17/work_ma/ww3_ounf_inla: Is a directory
> Warning: ww3_ounf_inla is different
> 
>    ****************************************************************************
>  ***  end of WAVEWATCH III compare tests with the same version of the code   ***
>    ****************************************************************************
> ```
> 
> Is this expected behavior?
> 
> After the grib programs run, we are getting:
> warning: empty layer generated
> and in fact if you convert the grib files to netcdf, they are all empty. This is likely a module file issue. We need to use jasper/2.0.25 for the grib files to be non-empty.

@JessicaMeixner-NOAA Done

JessicaMeixner-NOAA commented:
@aliabdolali are you ready for me to re-test and review again?

aliabdolali commented:
> @aliabdolali are you ready for me to re-test and review again?

There is one more comment by you to change: add cp rmp_*nc to run_test
I am working on it, then I will recheck the results.

Commit:
add copy rmp_src* to run_test instead of matrix.base

Commit:
Merge remote-tracking branch 'upstream/develop' into fb_ufs_regtests

Commit:
update ww3_multi.inp files for new fields introduced since UKMET coupling PR

aliabdolali commented:
@JessicaMeixner-NOAA 
There is no strange hs for ufs1.1 with a single grid. However, when I ran the ufs1.2 test (gfsv16), I noticed the sharp changes from time step to time step in hs field. I check the test with production/GFSv16 branch using run_test from this FB and it was ok.
So, we need to track and find when this happened. We need to find it immediately. This test shows how important such a test is. 
Could you review it again and let me know if you are ok?
I want to move on to the next PR.

JessicaMeixner-NOAA commented:
@aliabdolali I will review this again today. However, I don't think I'm going to be okay merging in a test that does not look okay.  I'll reply as soon as I have finished reviewing.

JessicaMeixner-NOAA commented:
@aliabdolali two questions: 
1. is there new output with the matrix output since changes have been made? 
2. Do you have a location of the ufs1.2 with the operations branch output I could see to compare? 
Thanks for reducing the resources significantly.

aliabdolali commented:
> @aliabdolali two questions:
> 
> 1. is there new output with the matrix output since changes have been made?
> 2. Do you have a location of the ufs1.2 with the operations branch output I could see to compare?
>    Thanks for reducing the resources significantly.

I haven't run the whole test again.
I deleted the production one.
I checked a commit before ukmet-coupling and it was still problematic,
So I checked a commit from August
_commit ca406df833c2af10cfeab8ca5343c27af6e8e8f9 (HEAD)
Author: Jessica Meixner <jessica.meixner@noaa.gov>
Date:   Mon Aug 3 23:09:14 2020 -0400_
and the results are reasonable.
/scratch2/COASTAL/coastal/save/Ali.Abdolali/Source2021/May30GFSv16/regtests/ww3_ufs1.2/work_a

Commit:
fix the missing "||" in cmplr.env

Commit:
Merge remote-tracking branch 'upstream/develop' into fb_ufs_regtests

Commit:
Merge remote-tracking branch 'upstream/develop' into fb_ufs_regtests

Commit:
add capablity to execute multiple ww3_prnc for multi setup

Commit:
update the run to 2021 with global coverage

Commit:
update ww3_multi*inp files for ww3_tp2.21 test

Commit:
Merge remote-tracking branch 'upstream/develop' into fb_ufs_regtests

Commit:
Merge remote-tracking branch 'upstream/develop' into fb_ufs_regtests

aliabdolali committed:
Update ww3_from_ftp.sh

aliabdolali commented:
@JessicaMeixner-NOAA 
I addressed the comments you made and I ran the full matrix again, with expected changes due to this development:
```
ww3_tp2.17
ww3_tp2.6
```

and known non-b4b tests:
```
mww3_test_03
mww3_test_07
ww3_tp2.10
ww3_tp2.15
```
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6683594/matrixDiff.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6683595/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6683596/matrixCompFull.txt)

I think this is ready to be merged and if you think we should wait for gfsv16 related bug, I can propose to exculde ww3_ufs1.2 (gfsv16) out and merge this one. In this way, we integrate the fixes and developments and leave the problematic one remained to be added to the bug fix for https://github.com/NOAA-EMC/WW3/issues/417

JessicaMeixner-NOAA commented:
@aliabdolali I will review again today.

JessicaMeixner-NOAA commented on regtests/bin/matrix_ncep:
```diff
@@ -63,15 +71,26 @@
+    modbacio='bacio/2.4.1'
+    modg2='g2/3.4.1'
+    modw3nco='w3nco/2.4.1'
+    modesmf='esmf/8_1_0_beta_snapshot_47'
```
Now that esmf/8_1_1  is available, we could use that?  We can also update that later.

aliabdolali replied:
I'll update it.

aliabdolali replied:
I changed it and tested ww3_multi_esmf build with it.

Commit:
update esmf version to 8_1_1

Commit:
Merge remote-tracking branch 'upstream/develop' into fb_ufs_regtests

JessicaMeixner-NOAA commented:
In the PR description, I could not tell what would make tp2.6 and tp2.17 change?  Which updates cause those changes exactly? 


There is the issue with the new ufs 1.1 and 1.2 test which has been document in issue #417,  and shown to be not a cause of this addition but an existing bug and therefore I think this can be merged. One comment on the wind input for ufs 1.2 is that the ice fields are blocking the u,v values. Therefore if you use a different ice input or if it varies over time and does not match the wind input itself, someone might have issues.   As long as that is known and thought to be acceptable, we can move forward with this PR.

aliabdolali commented:
@JessicaMeixner-NOAA 

- For ww3_tp2.17, I changed the coordinates of point outputs. all of them were previously on the boundary nodes, now they are within the domain. I'll add a note in the PR description.

- For the ww3_tp2.6 in matrixCompFull.txt:
ww3_tp2.6/./work_pdlib                     (0 files differ)
so it is ok.

- I'll add a note of the ufs1.2 issue (#417)

JessicaMeixner-NOAA commented on regtests/ww3_ufs1.1/info:
```diff
@@ -0,0 +1,60 @@
+#                                                                           #
+# Grids:                                                                    #
+# MODEL: glo_1deg with 1 deg resolution from 85S to 85N.                    #
+# INPUT: glo_30m.The forcing including ice concentration, wind speed (u10)  #
```
this is now 15mext 

Also, I made this file: /scratch2/NCEPDEV/climate/Jessica.Meixner/forinputww3ufstest/gfs.20210401/00/atmos/wind.2021040100.nc which is from gfs output and has hourly wind for 2021040100, could we use this instead of the wind that is blocked out by ice?

Commit:
change the wind input to hourly gfs/atmos with no ice mask and update both ufs tests

JessicaMeixner-NOAA commented:
@aliabdolali thanks for updating the wind.  The new ufs regression tests run except for the issue which is documented in #417.  The info file for both of those tests are now out of date with the changes that have occurred since originally added.  After those are updated and any last regtests are run, this should be ready to be merged.

Commit:
update info for ufs tests and rename aux to tools in model/.gitignore

aliabdolali commented:
I ran the full matrix again, with expected changes in the inputs for:
_ww3_tp2.17_
and known non-b4b tests:

mww3_test_03
mww3_test_07
ww3_tp2.10
ww3_tp2.15

[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6712331/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6712333/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6712334/matrixDiff.txt)
